### PR TITLE
Child school relation impl

### PR DESCRIPTION
**Pull request by TheSlimvReal**
see issue: #216 

This branch introduces the ChildSchoolRelations.
This relations descripte a child visiting a school for a certain time period and class.

This relation should be used instead of giving each child a schoolId which this child is visting at the moment.

I removed the schoolId and schoolClass tags from the Child class and introduced the new ChildWithRelation class. This class gets a child and relation object and performs exactly like the old child object (with schoolId and schoolClass). This could maybe be refactored sometimes.

I also generated database indices and queries to fetch this relations effectivly.

### Visible/Frontend Changes
- [x] 
- [ ] 

### Architectural/Backend Changes
- [x] 
- [ ]

### CODE REVIEW

File: src/app/children/childWithRelation.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import {Child} from './child';
+import {ChildSchoolRelation} from './childSchoolRelation';
+
+export class ChildWithRelation extends Child {
```
Comment (sleidig) [inline src/app/children/childWithRelation.ts]: Why do we need a separate class ChildWithRelation? I think it would be much clearer if we just modify the existing Child class (and leave the schoolRelation undefined where necessary). Or am I missing something?

File: src/app/children/children.service.ts [hunk 2]
```diff
@@ -80,7 +99,72 @@ export class ChildrenService {
...
+          map: `(doc) => {
+            if (!doc._id.startsWith("${ChildSchoolRelation.ENTITY_TYPE}")) return;
+            if (doc.end) return;
+            emit(doc.schoolId);
+            }`
+        },
+        by_date: {
+          map: `(doc) => {
+            if (!doc._id.startsWith("${ChildSchoolRelation.ENTITY_TYPE}")) return;
+            emit(doc.childId + (new Date(doc.start)).getTime());
```
Comment (sleidig) [inline src/app/children/children.service.ts]: maybe for readability during debugging add a separator character: `emit(doc.childId + '_' + (new Date(doc.start)).getTime());`

File: src/app/children/children.service.ts [hunk 3]
```diff
@@ -198,4 +274,23 @@ export class ChildrenService {
...
+    return this.queryLatestRelation(childId)
+      .then(relation => {
+        if (relation) {
+         return this.entityMapper.load<School>(School, relation.schoolId)
+        }
+        return null;
+      });
+  }
+
+  async getViewableSchools(childId: string): Promise<SchoolWithRelation[]> {
```
Comment (sleidig) [inline src/app/children/children.service.ts]: In which sense are the Schools returned by `getViewableSchools()` "viewable"? Maybe we could name this in a more obvious way? The code looks like it loads all schools of a child?

File: src/app/children/view-schools-component/edit-school-dialog/edit-school-dialog.component.html [hunk 4]
```diff
@@ -10,7 +10,7 @@ <h2 mat-dialog-title>{{creating ? 'Add' : 'Edit'}} school for {{child.name}}</h2
 </mat-form-field>
 
 <mat-form-field>
-  <input matInput [(ngModel)]='childSchoolRelation.schoolClass' placeholder='Class' title='class' type='text'>
+  <input matInput [(ngModel)]='childSchoolRelation.class' placeholder='Class' title='class' type='text'>
```
Comment (sleidig) [inline src/app/children/view-schools-component/edit-school-dialog/edit-school-dialog.component.html]: `class` is or may be a protected keyword of TypeScript, that's why I renamed the attribute to `schoolClass` initially.

File: src/app/children/view-schools-component/view-schools.component.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+import {ChildrenService} from '../children.service';
+import {SchoolWithRelation} from '../../schools/schoolWithRelation';
+
+@Component({
+  selector: 'app-view-schools-component',
+  templateUrl: './view-schools.component.html',
+  styleUrls: ['./view-schools.component.scss']
+})
+
+export class ViewSchoolsComponent implements OnInit, OnChanges {
```
Comment (sleidig) [inline src/app/children/view-schools-component/view-schools.component.ts]: Could this use our `EntitySubrecordComponent`? That might avoid writing some of this code that's similar in other components.

File: src/app/entity/PassableEntityConstructor.ts [hunk 6]
```diff
@@ -0,0 +1,3 @@
+import {Entity} from './entity';
+
+export type PassableEntityConstructor<T extends Entity> = new(id: string) => T;
```
Comment (sleidig) [inline src/app/entity/PassableEntityConstructor.ts]: Good idea to define a proper type for this rather than always having to search for complicated type definition!

I would simply name it `EntityConstructor` though - we don't have a non-passable one anyway.

As this is just a one liner and closely related, maybe we add it to the entity.ts file rather than in a separate one?

File: src/app/schools/school.ts [hunk 7]
```diff
@@ -1,24 +1,23 @@
...
 
 export class School extends Entity {
-  static ENTITY_TYPE = 'School';
+  public static ENTITY_TYPE = 'School';
 
-  name = '';
-  medium = '';
-  privateSchool: boolean;
-  academicBoard = '';
+  public name = '';
```
Comment (sleidig) [inline src/app/schools/school.ts]: attributes are "public" by default. Any reason to add it here explicitly for some of the attributes?

File: src/app/schools/schoolWithRelation.ts [hunk 8]
```diff
@@ -0,0 +1,32 @@
+import {ChildSchoolRelation} from '../children/childSchoolRelation';
+import {School} from './school';
+
+export class SchoolWithRelation {
```
Comment (sleidig) [inline src/app/schools/schoolWithRelation.ts]: Same discussion point as for `ChildWithRelation`: It would be easier to understand if there is just a `School` class that integrates this directly in my opinion.

File: src/app/schools/schools-list/schools-list.component.html [hunk 9]
```diff
@@ -35,7 +35,7 @@ <h1>Schools List</h1>
 </mat-expansion-panel>
 
 <div *ngIf="schoolDataSource">
-  <table mat-table [dataSource]="schoolDataSource" matSort class="mat-elevation-z8 table-list" width='100%'>
+  <mat-table mat-table [dataSource]="schoolDataSource" matSort class="mat-elevation-z8 table-list" width='100%'>
```
Comment (sleidig) [inline src/app/schools/schools-list/schools-list.component.html]: Angular Material documentation recommends to use `<table mat-table>` (https://material.angular.io/components/table). I think that has some advantages over `<mat-table>` regarding how browsers handle it. Any particular reason to change it here?

File: src/app/ui/search/search.component.ts [hunk 10]
```diff
@@ -1,8 +1,8 @@
 import { Component, OnInit } from '@angular/core';
 import {Database} from '../../database/database';
 import {Child} from '../../children/child';
-import {School} from '../../schools/school';
 import {Entity} from '../../entity/entity';
+import {School} from '../../schools/school';
```
Comment (sleidig) [inline src/app/ui/search/search.component.ts]: if you notice it, try to avoid committing such small, unrelated changes. The less changes, the easier is the review and merging :-)

File: src/app/children/childWithRelation.ts (hunk 1 above)
Comment (TheSlimvReal) [reply src/app/children/childWithRelation.ts]: When we want to move this to the original child class, we need some functionality in the entity mapper that prevents the additional information (like class and school) to be saved to the db. Maybe we could say the nested objects are not stored at all or we only store attributes without a leading "_". If you want we can have a call about that topic this week.

File: src/app/children/children.service.ts (hunk 3 above)
Comment (TheSlimvReal) [reply src/app/children/children.service.ts]: Maybe you could call it just getSchoolWithRelation. It's viewable in that sense, that you can display the name of the related school instead of just the schoolId which is saved in the relation.

File: src/app/children/view-schools-component/edit-school-dialog/edit-school-dialog.component.html (hunk 4 above)
Comment (TheSlimvReal) [reply src/app/children/view-schools-component/edit-school-dialog/edit-school-dialog.component.html]: Oh sorry I dind't know that. I was not sure, what the current version of it was.

File: src/app/children/childWithRelation.ts (hunk 1 above)
Comment (sleidig) [reply src/app/children/childWithRelation.ts]: That's true. The introduction of an EntitySchema (see #233) would solve this by making the decision what is saved to the database more explicit.

File: src/app/children/view-schools-component/view-schools.component.ts (hunk 5 above)
Comment (TheSlimvReal) [reply src/app/children/view-schools-component/view-schools.component.ts]: I will talk to @codingfabi and @christophscheuing what they think about it.

File: src/app/children/view-schools-component/edit-school-dialog/edit-school-dialog.component.html (hunk 4 above)
Comment (sleidig) [reply src/app/children/view-schools-component/edit-school-dialog/edit-school-dialog.component.html]: Funny enough, I never saw any issues with using `class` but it's probably better to avoid it.

File: src/app/schools/school.ts (hunk 7 above)
Comment (TheSlimvReal) [reply src/app/schools/school.ts]: Your right, I am more used to do it that way but I will change it back to keep it in line with the project-style.

Comment (sleidig) [reply src/app/schools/school.ts]: Maybe also discuss it in the next meeting. If most people feel it's clearer if we explicitly add "public", we can also do that across all classes.

File: src/app/children/childWithRelation.ts (hunk 1 above)
Comment (TheSlimvReal) [reply src/app/children/childWithRelation.ts]: So what do you recommend? Should we leave it like this until you are done with #233 or should we implement something temporary?

File: src/app/children/view-schools-component/view-schools.component.ts (hunk 5 above)
Comment (TheSlimvReal) [reply src/app/children/view-schools-component/view-schools.component.ts]: Turns out @christophscheuing is already working on it (#207). But I think it would be good to merge these changes soon because I also edited this component.

Comment (sleidig) [reply src/app/children/view-schools-component/view-schools.component.ts]: ah, nice. Maybe you guys can discuss directly which change should go in first and in which way.

File: src/app/children/childWithRelation.ts (hunk 1 above)
Comment (sleidig) [reply src/app/children/childWithRelation.ts]: Might be that #233 leads to a couple of interface changes that touch many files, which then creates complicated merges ... So I would prefer getting this done independent of it first.

What if we just add class and school to the `Child` class anyway - and accept that for a brief period that information will be written to the database unintentionally until the Schema system is in master? And as soon as the option is available, we can exclude those fields from the database schema.
At first glance, I don't see that causing major problems. What do you think?

Comment (TheSlimvReal) [reply src/app/children/childWithRelation.ts]: When this is anyway just a temporary solution, can't we stick to the ChildWithRelation and SchoolWithRelation for now? I think it shows very clearly where changes will happen when #233 is ready and it would be good @christophscheuing to continue with his task.
